### PR TITLE
Bash Scripting Blog Post

### PR DESCRIPTION
**Pull request by JoshArmi**

### CODE REVIEW

File: _posts/2018-08-10-better-bash-in-5-minutes.markdown [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        docker container stop container1
+        docker container rm container1
+    }
+    trap cleanup SIGINT SIGTERM EXIT
+```
+
+## Learning when Bash is no longer the tool for the job
+Every time I extend a bash script, I always consider the trade off between adding complexity to this script vs. potentially rewriting in Python or another language. Bash has it's advantages, but it's worth considering for critical scripts to write them in something easier to maintain and extend.
+
+## Furter Reading
```
Comment (paulmech) [inline _posts/2018-08-10-better-bash-in-5-minutes.markdown]: Further

File: _posts/2018-08-10-better-bash-in-5-minutes.markdown [hunk 2]
```diff
@@ -0,0 +1,71 @@
+---
+layout: post
+title:  "Better Bash Scripting in 5 Minutes or Less"
+date:   2018-08-10
+categories: bash linux unix mac osx script shell
+author: Josh Armitage
+image: img/bash-logo.png
+---
+
+## Shebang `#!/bin/bash`
```
Comment (Hamo69) [inline _posts/2018-08-10-better-bash-in-5-minutes.markdown]: Do we need to explain Shebang any more?

File: _posts/2018-08-10-better-bash-in-5-minutes.markdown [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+This should be added unless you're checking for the existence of a variable, so if you reference any variables by mistake or typo the script will instantly fail.
+
+## Add trace logging `set -x`
+During the debugging stages it can be helpful to see what is being run, with this set it will print out the commands in full as they're run with all the variables resolved.
+
+## Piping Carefully `set -o pipefail`
+By default if you pipe in bash the return code is the return code of the last statement in the pipe. After setting this, if any command returns a non-zero return code the pipe will fail.
+
+## Simple Command Output Checking
+This pattern for checking output while not perfect is simple and covers the vast majority of cases. The below statement idempotently stops and removes a Docker container.
```
Comment (Hamo69) [inline _posts/2018-08-10-better-bash-in-5-minutes.markdown]: do we need to explain idempotence?

File: _posts/2018-08-10-better-bash-in-5-minutes.markdown [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    do
+        docker run ${CONTAINER}
+    done
+```
+
+## Idempotency
+When the scripts are being committed to source control and shared amongst a team it's worth spending the time making sure that all commands are run idempotently but will still fail as expected.
+
+## Reassign Parameters to Named Variables
+Parameters passed to the script come through as $1, $2, etc. Especially as the bash file grows in size the meaning of the variables gets lost. Good practice is to reassign them at the top of the file.
```
Comment (Hamo69) [inline _posts/2018-08-10-better-bash-in-5-minutes.markdown]: 👍

File: _posts/2018-08-10-better-bash-in-5-minutes.markdown (hunk 3 above)
Comment (JoshArmi) [reply _posts/2018-08-10-better-bash-in-5-minutes.markdown]: You'd be surprised ;)

**Pull request merged**